src/nvim/extmark.c

```diff
@@ -67,18 +67,18 @@ void extmark_set(buf_T *buf, uint32_t ns_id, uint32_t *idp, int row, colnr_T col
       } else {
         assert(marktree_itr_valid(itr));
         if (old_mark.pos.row == row && old_mark.pos.col == col) {
+          // not paired: we can revise in place
           if (mt_decor_any(old_mark)) {
+            mt_itr_rawkey(itr).flags &= (uint16_t) ~MT_FLAG_DECOR_SIGNTEXT;
             buf_decor_remove(buf, row, row, mt_decor(old_mark), true);
           }
-
-          // not paired: we can revise in place
           mt_itr_rawkey(itr).flags &= (uint16_t) ~MT_FLAG_EXTERNAL_MASK;
           mt_itr_rawkey(itr).flags |= flags;
           mt_itr_rawkey(itr).decor_data = decor.data;
           goto revised;
         }
-        buf_decor_remove(buf, old_mark.pos.row, old_mark.pos.row, mt_decor(old_mark), true);
         marktree_del_itr(buf->b_marktree, itr, false);
+        buf_decor_remove(buf, old_mark.pos.row, old_mark.pos.row, mt_decor(old_mark), true);
```

Comment:
Is this change necessary?

Reply:
Hmm since the counting is now delayed to the start of a redraw it no longer is. Can revert I guess but I think it's more correct in this order in case `buf_decor_remove()` ever handles other side effects which expect the mark to already be removed from the marktree.

Reply:
The other line only deletes an iterator. It doesn't change the marktree.

Reply:
Uhm I don't think that's true. `marktree_del_itr()` removes the key pointed to by the iterator?